src/services/db/GitFileSystemService.ts

```diff
@@ -363,24 +392,45 @@ export default class GitFileSystemService {
         )
       }
 
-      return this.ensureCorrectBranch(repoName).andThen(() =>
-        ResultAsync.fromPromise(
-          isForce
-            ? this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).push(["--force"])
-            : this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).push(),
-          (error) => {
-            logger.error(`Error when pushing ${repoName}: ${error}`)
+      return this.ensureCorrectBranch(repoName)
+        .andThen(() =>
+          ResultAsync.fromPromise(
+            isForce
+              ? this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).push(["--force"])
+              : this.git.cwd(`${EFS_VOL_PATH}/${repoName}`).push(),
+            (error) => {
+              logger.error(`Error when pushing ${repoName}: ${error}`)
 
-            if (error instanceof GitError) {
-              return new GitFileSystemError(
-                "Unable to push latest changes of repo"
-              )
+              if (error instanceof GitError) {
+                return new GitFileSystemError(
+                  "Unable to push latest changes of repo"
+                )
+              }
+
+              return new GitFileSystemError("An unknown error occurred")
             }
+          )
+        )
+        .orElse(() =>
+          // Retry push once
+          ResultAsync.fromPromise(
```

Comment:
Hmm I think it is somewhat expected that pushes may fail due to concurrency, as pushes are done asynchronously for each repo modification operation (i.e. push is done while a response is returned to the user). Since our filesystem operations are running very quickly, if we have an instance of a delete-then-create (i.e. renaming pages), it is easy to get the push operation from delete conflicting with the push operation from create. The create operation does not wait for the push operation to complete, and thus it makes it more likely for a concurrency conflict to occur.

I think we can try out with the exponential backoff method as you suggested.

Reply:
yea also the check for git index.lock to see if any other push etc is happening

Reply:
@dcshzj the exponential backoff can be an enhancement later as well

Reply:
Ok sure, we can do those as part of IS-434.